ethcore/builtin/src/lib.rs

```diff
@@ -420,10 +420,14 @@ impl Implementation for EcRecover {
 
 		let s = Signature::from_rsv(&r, &s, bit);
 		if s.is_valid() {
-			if let Ok(p) = ec_recover(&s, &hash) {
+			let recovery_message = ZeroesAllowedMessage(hash);
+			if let Ok(p) = recover_allowing_all_zero_message(&s, recovery_message) {
 				let r = keccak(p);
 				output.write(0, &[0; 12]);
 				output.write(12, &r.as_bytes()[12..]);
+			} else {
+				// todo[dvdplm] do the right thing with errors here
+				warn!("Public key recovery from signed message failed");
```

Comment:
The error type is just `&'static str`, so should be trivial to fix.

Maybe add another error if the signature is invalid too.

Reply:
So my doubt here is this: what does it mean when `ecrecover` fails? Is it an error for `openethereum` to handle or should it be handled in the calling code, at the Solidity level? Like, when a contract calls the builtin and it fails then `output` will be left untouched (hopefully all zeroes?), and maybe that's how the caller knows the recovery fails?
If we return an error here, what happens to the calling code? Does it `REVERT`?

@q9f do you know what the right thing to do here is? Or @adria0 ? @sorpaas ?

On the `openethereum` level, not having any log or error returned here made finding the bug quite tricky, so from that POV I'd want to do _something_. When I synced the various chains, this warning message is printed every so often so it is definitely "normal".

Reply:
It must return zeros on error according

Reply:
@dvdplm Yeah this warning can be a DoS vector because inputs are controlled by users.

Attacker: Pay some thousands of gas and I can get a warning message printed in every client, each time.